### ISSUE TITLE
How can I  draw a rect between the window bg color and the frame content of a TreeNode item?

### ISSUE THREAD

**Issue by khiner** [ocornut/imgui#5334]
**Version/Branch of Dear ImGui:**

Version: 1.88WIP
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: `imgui_impl_sdl.cpp` + `imgui_impl_opengl3.cpp`
Operating System: MacOS

**My Issue/Question:**

I would like to draw a filled rectangle _behind_ a `TreeNode` item's contents (its text & expand/collapse icon), but _above_ the background paint of its parent window. If I use `GetBackgroundDrawList()->AddRectFilled`, the rect gets added _under the window background_. I want to render between the window background and the `TreeNode` (right behind the node's contents).

The way I'm currently solving this is to set my `ImGuiCol_WindowBg` to transparent, but of course this isn't a real solution.

Here is what I'm doing currently:

```cpp
if (ImGui::TreeNode(label)) {
    ImGui::Text("%s : %s", name.c_str(), value.dump().c_str());
    ImGui::GetBackgroundDrawList()->AddRectFilled(row_min, row_max, color, 0.0f, ImDrawFlags_None);
    ImGui::TreePop();
}
```

If I instead use `GetForegroundDrawList` or `GetWindowDrawList`, the rect gets drawn on top of the TreeNode's contents (text & expand/collapse icon).

I also want to keep the existing hovered/selected background rect color behavior, and draw my custom rect on top of that selected/hovered frame rect as well.

Thank you!

**Comment by ocornut** [comment]
I would say the best solution would depend on what and why you are drawing (criteria for drawing that filled background, ratio of items that will draw it).

> behind a TreeNode item's contents (its text & expand/collapse icon)

If it is actually just behind the node text and icon, then you can simply calculate its bounding box manually ahead of time and draw before?

If it is behind the full content inside the node that's trickier.

You can use a `ImDrawListSplitter` to draw things out of order (switch to channel 1 when submitting tree nodes, then switch to channel 0 to draw your stuff). That would work but if you consistently switch channels for every tree node it might be a bit suboptimal. Or maybe ok.

If many of your items need a bg you might always draw a rectangle or arbitrary shape BEFORE the TreeNode, without rounding, so the vertices are reserved. You record the vertex location before drawing the rectangle. You know a regular rectangle takes 4 vertices (two triangles) and you can move the vertices at the end of the node contents once you know the total contents size. If some items don't need the filled background, clear those vertices by overwriting alpha to 0 and copying the first vertice into others to create two zero-sized triangles.

Use Metrics->DrawList to investigate how vertices are emitted for rectangles.

**Comment by khiner** [comment]
Thanks for the quick response.

My use case: I'm rendering my main application JSON state to an ImGui tree for debugging purposes. I am using background rects in a tree to represent the number of times any leaf node (ultimately representing some potentially nested member of my root state `struct`) has been updated, relative to other state items. So essentially using the background colors of each tree node as a relative update-frequency histogram, to see where most of my app's update activity is. Hope that makes sense.

Here is a screenshot of this done in a proper, flat histogram, using ImPlot:

<img width="653" alt="Screen Shot 2022-05-22 at 10 53 14 AM" src="https://user-images.githubusercontent.com/685394/169708863-d7ba5069-1536-4484-9b49-65d9dde70021.png">

and here is my WIP tree-version, showing a small subset of the state, hopefully you get the idea:

<img width="653" alt="Screen Shot 2022-05-22 at 10 53 31 AM" src="https://user-images.githubusercontent.com/685394/169708883-64ce0ead-3460-484d-8466-71cb30fc5f73.png">


> If it is actually just behind the node text and icon, then you can simply calculate its bounding box manually ahead of time and draw before?

This sounds like exactly what I'm looking for. Was the rest of your answer mostly explaining how to actually go about this, or was that all assuming "If it is behind the full content inside the node that's trickier."?

**Comment by ocornut** [comment]
This sounds like exactly what I'm looking for. Can you say more about how to accomplish exactly this? 

Use `GetWindowDrawList()->AddRectFilled()`.
The upper-left position you know is your cursor position (`GetCursorScreenPos()`).
The height of a tree node is `GetFontSize()`. Looks like there's nothing else you need?

Second part was that all assuming "If it is behind the full content inside the node that's trickier."?

**Comment by khiner** [comment]
Thank you, I did try this but using `GetWindowDrawList` is rendering the rect on top of the text:

if (ImGui::TreeNode(label)) {
    ImGui::GetWindowDrawList()->AddRectFilled(row_min, row_max, color, 0.0f, ImDrawFlags_None);
    ImGui::Text("%s : %s", name.c_str(), value.dump().c_str());
    ImGui::TreePop();
}

**Comment by ocornut** [comment]
Well you need to render before the TreeNode…

**Comment by khiner** [comment]
I figured out my issue. I was using `GetItemRectMin()`/`IGetItemRectMax()` to calculate the bound of my rect, instead of the `GetCursorScreenPos()`/`GetFontSize()` like you suggested. This was working if I rendered the rect after the tree node, and in the background, but when rendering before the tree node it was rendering in the _previous_ node's position, on top of the node contents.

This behavior all makes sense to me now (`GetItemRect(Min|Max)` used to get the position of the last item, `GetCursorScreenPos` to get the upcoming item pos & using content-specific calculations (`GetFontSize` in this case) for the max pos).

I was super confused why `GetWindowDrawList` wasn't doing what I expected, but now I see it's user error!

Thanks! Closing.
